### DISCUSSION TITLE
task 8 reshape value error

### DISCUSSION THREAD

**Discussion by tdeng455** [Part-II-Computational-Physics/.github#46] [General]
Hi I am trying to check that I have defined my linear operator correctly but I am not sure how to call the LinearOperator function within the definition of my LinearHamiltonian
I have made an attempt that looks like:
```py
def LinearHamiltonian(N, h=1): 
    H = spsl.LinearOperator(shape=(2**N,2**N), matvec=ActWithH(psi,N))
    return H

print(LinearHamiltonian(3,1))

```
and I get the error calling back to task 7 which I used the notes to write
the error is:
ValueError: cannot reshape array of size 16 into shape (16,16)

```py
def ActWithH(psi, N, h=1.):
    
    Hloc2 = LocalHamiltonian(h)
    D = Hloc2.shape[0]
    psiOut = np.zeros(psi.size)
    for k in range(N-1):

        psiOut += np.tensordot(Hloc2.reshape(D**2,D**2),psi.reshape(D**k,D**2,D**(N-2-k)), axes=[[1],[1]]).transpose(1,0,2).reshape(D**N)

    #boundrary term
    psiOut += np.tensordot(Hloc2.reshape(D,D,D,D),psi.reshape(D,D**(N-2),D), axes=[[2,3],[2,0]]).transpose(1,2,0).reshape(D**N)
    
    return psiOut
```
which seems correct to me...

**Comment by JordanEmil** [comment]
I'm having the same issue.

**Reply by tdeng455** [reply]
I have changed `D` defined in task 7 to `D=2` and redefined `LinearHamiltonian` to only be a function of `psi` defined as
```py

def LinearHamiltonian(psi):
    
    H = ActWithH(psi,N,h)

    return H

```
this seems to have worked

**Comment by JNHallen** [comment]
Hi!  

The issue you are having is that the notes defined Hloc as a (2,2,2,2) array, which means that Hloc2.shape[0] = 2. We have defined it in the (4, 4) form instead, so Hloc.shape[0] = 4.  The D you want to use would still be 2.
